File: swift/Sources/CantonKit/LedgerUpdate.swift
import CantonLedgerAPI

/// One element of the ledger update stream. Every update carries the
/// participant-local ``offset`` that totally orders it; persist the offset of
/// the last update you processed and pass it as
/// ``UpdateSubscription/beginExclusive`` to resume without gaps or duplicates.
public enum LedgerUpdate: Sendable {
    /// A committed Daml transaction.
    case transaction(Com_Daml_Ledger_Api_V2_Transaction)

    /// A contract moving between synchronizers.
    case reassignment(Com_Daml_Ledger_Api_V2_Reassignment)

    /// A party (de)activation on the participant.
    case topologyTransaction(Com_Daml_Ledger_Api_V2_TopologyTransaction)

    /// A cursor keep-alive: no matching event, but the resume offset advanced.
    case checkpoint(offset: Int64)

    public var offset: Int64 {
        switch self {
        case .transaction(let transaction): transaction.offset
        case .reassignment(let reassignment): reassignment.offset
        case .topologyTransaction(let topology): topology.offset
        case .checkpoint(let offset): offset
        }
    }

    init?(_ response: Com_Daml_Ledger_Api_V2_GetUpdatesResponse) {
        switch response.update {
        case .transaction(let transaction):
            self = .transaction(transaction)
        case .reassignment(let reassignment):
            self = .reassignment(reassignment)
        case .topologyTransaction(let topology):
            self = .topologyTransaction(topology)
        case .offsetCheckpoint(let checkpoint):
            self = .checkpoint(offset: checkpoint.offset)
        case .none:
            return nil // unknown future update kind; skip
        }
    }
}

File: kotlin/canton-sdk/src/main/kotlin/io/github/vsima/canton/LedgerUpdate.kt
package io.github.vsima.canton

import com.daml.ledger.api.v2.ReassignmentOuterClass
import com.daml.ledger.api.v2.TopologyTransactionOuterClass
import com.daml.ledger.api.v2.TransactionOuterClass
import com.daml.ledger.api.v2.UpdateServiceOuterClass.GetUpdatesResponse

/**
 * One element of the ledger update stream. Every update carries the
 * participant-local [offset] that totally orders it; persist the offset of
 * the last update you processed and pass it as
 * [UpdateSubscription.beginExclusive] to resume without gaps or duplicates.
 */
public sealed interface LedgerUpdate {
    public val offset: Long

    /** A committed Daml transaction. */
    public data class Transaction(
        val transaction: TransactionOuterClass.Transaction,
    ) : LedgerUpdate {
        override val offset: Long get() = transaction.offset
    }

    /** A contract moving between synchronizers. */
    public data class Reassignment(
        val reassignment: ReassignmentOuterClass.Reassignment,
    ) : LedgerUpdate {
        override val offset: Long get() = reassignment.offset
    }

    /** A party (de)activation on the participant. */
    public data class TopologyTransaction(
        val topologyTransaction: TopologyTransactionOuterClass.TopologyTransaction,
    ) : LedgerUpdate {
        override val offset: Long get() = topologyTransaction.offset
    }

    /**
     * A cursor keep-alive: no matching event, but the resume offset advanced.
     */
    public data class Checkpoint(override val offset: Long) : LedgerUpdate

    public companion object {
        internal fun from(response: GetUpdatesResponse): LedgerUpdate? = when {
            response.hasTransaction() -> Transaction(response.transaction)
            response.hasReassignment() -> Reassignment(response.reassignment)
            response.hasTopologyTransaction() -> TopologyTransaction(response.topologyTransaction)
            response.hasOffsetCheckpoint() -> Checkpoint(response.offsetCheckpoint.offset)
            else -> null // unknown future update kind; skip
        }
    }
}
